controller/src/main/java/io/pravega/controller/store/stream/records/StreamTruncationRecord.java

```diff
@@ -81,17 +82,44 @@
     private final boolean updating;
 
     @Builder
-    public StreamTruncationRecord(Map<Long, Long> streamCut, Map<StreamSegmentRecord, Integer> span,
-                                  Set<Long> deletedSegments, Set<Long> toDelete, long sizeTill, boolean updating) {
-        this.streamCut = ImmutableMap.copyOf(streamCut);
-        this.span = ImmutableMap.copyOf(span);
-        this.deletedSegments = ImmutableSet.copyOf(deletedSegments);
-        this.toDelete = ImmutableSet.copyOf(toDelete);
+    /**
+     * This is a private constructor that is only directly used by the builder during the deserialization. 
+     * The deserialization passes @param copyCollections as false so that we do not make an immutable copy of the collection
+     * for the collection passed to the constructor via deserialization. 
+     *
+     * The all other constructors, the value of copyCollections flag is true and we make an immutable collection copy of 
+     * the supplied collection. 
+     * All getters of this class that return a collection always wrap them under Collections.unmodifiableCollection so that
+     * no one can change the data object from outside.  
+     */
+    private StreamTruncationRecord(Map<Long, Long> streamCut, Map<StreamSegmentRecord, Integer> span,
+                                  Set<Long> deletedSegments, Set<Long> toDelete, long sizeTill, boolean updating, boolean copyCollections) {
+        this.streamCut = copyCollections ? ImmutableMap.copyOf(streamCut) : streamCut;
+        this.span = copyCollections ? ImmutableMap.copyOf(span) : span;
+        this.deletedSegments = copyCollections ? ImmutableSet.copyOf(deletedSegments) : deletedSegments;
+        this.toDelete = copyCollections ? ImmutableSet.copyOf(toDelete) : toDelete;
         this.sizeTill = sizeTill;
         this.updating = updating;
         this.spanEpochLow = span.values().stream().min(Comparator.naturalOrder()).orElse(Integer.MIN_VALUE);
         this.spanEpochHigh = span.values().stream().max(Comparator.naturalOrder()).orElse(Integer.MIN_VALUE);
     }
+    
+    public StreamTruncationRecord(Map<Long, Long> streamCut, Map<StreamSegmentRecord, Integer> span,
+                                  Set<Long> deletedSegments, Set<Long> toDelete, long sizeTill, boolean updating) {
+        this(streamCut, span, deletedSegments, toDelete, sizeTill, updating, true);
+    }
+
+    public Map<Long, Long> getStreamCut() {
+        return Collections.unmodifiableMap(streamCut);
```

Comment:
If the value passed into the constructor were always copied to an immutable map this would not be needed.

Reply:
yes, but thats exactly what we are trying to avoid. creation of duplicate collection nodes. please refer to the comment above where i have explained the rationale and approach.